components/support/nimbus-fml/src/util/loaders.rs

```diff
@@ -0,0 +1,289 @@
+            .build()?;
+
+        Ok(Self {
+            cache_dir,
+            fetch_client: http_client,
+
+            config,
+        })
+    }
+
+    pub fn default() -> Result<Self> {
+        let cwd = std::env::current_dir()?;
+        let cache_path = cwd.join("build/app/fml-cache");
+        Self::new(cache_path, Default::default())
+    }
+
+    /// This loads a text file from disk or the network.
+    ///
+    /// If it's coming from the network, then cache the file to disk (based on the URL).
+    ///
+    /// We don't worry about cache invalidation, because a clean build should blow the cache
+    /// away.
+    pub fn read_to_string(&self, file: &FilePath) -> Result<String> {
+        Ok(match file {
+            FilePath::Local(path) => std::fs::read_to_string(path)?,
+            FilePath::Remote(url) => self.fetch_and_cache(url)?,
+        })
+    }
+
+    fn fetch_and_cache(&self, url: &Url) -> Result<String> {
+        let path_buf = self.create_cache_path_buf(url);
+        Ok(if path_buf.exists() {
+            std::fs::read_to_string(path_buf)?
+        } else {
+            let res = self.fetch_client.get(url.clone()).send()?;
+            let text = res.text()?;
+
+            std::fs::write(path_buf, &text)?;
+            text
+        })
+    }
+
+    fn create_cache_path_buf(&self, url: &Url) -> PathBuf {
+        let mut hasher = DefaultHasher::new();
+        url.hash(&mut hasher);
+        let checksum = hasher.finish();
+        let filename = match url.path_segments() {
+            Some(segments) => segments.last().unwrap_or("unknown.txt"),
+            None => "unknown.txt",
+        };
+        let filename = format!("{:x}_{}", (checksum & 0x000000000000FFFF) as u16, filename,);
+        self.cache_dir.join(filename)
+    }
+
+    /// Joins a path to a string, to make a new path.
+    ///
+    /// We want to be able to support local and remote files.
+    /// We also want to be able to support a configurable short cut format.
+    /// Following a pattern common in other package managers, `@XXXX/YYYY`
+    /// is used as short hand for the main branch in github repos.
+    pub fn join(&self, base: &FilePath, f: &str) -> Result<FilePath> {
+        if f.starts_with('@') {
+            let f = f.replacen('@', "", 1);
+            let parts = f.splitn(3, '/').collect::<Vec<&str>>();
+            match parts.as_slice() {
+                [user, repo, path] => {
+                    let repo = self.lookup_repo_path(user, repo);
+                    let f = format!("{}/{}", repo, path);
+                    base.join(&f)
+                }
+                _ => Err(FMLError::InvalidPath(format!(
+                    "'{}' needs to include a username, a repo and a filepath",
+                    f
+                ))),
+            }
+        } else {
+            base.join(f)
+        }
+    }
+
+    fn lookup_repo_path(&self, user: &str, repo: &str) -> String {
+        let default = "main".to_string();
```

Comment:
think `gecko-dev` still uses `master` as its default branch on the github mirror, so might need to account for that at some point